Types don't resolve when using `"moduleResolution": "nodenext"`

### Reproduction

```json
{
  "name": "byethrow-reproduction",
  "type": "module",
  "main": "index.ts",
  "scripts": {
    "compile": "tsc --noEmit"
  },
  "devDependencies": {
    "@tsconfig/node24": "^24.0.1",
    "typescript": "^5.9.2"
  },
  "dependencies": {
    "@praha/byethrow": "^0.7.0"
  }
}
```

```json
{
  "extends": "@tsconfig/node24/tsconfig.json",
  "include": ["./index.ts"]
}
```

```typescript
// index.ts
import { Result } from "@praha/byethrow";

const result = Result.succeed(4);

type R = Result.InferSuccess<typeof result>;

type Assert<T extends true> = T;
type IsEqual<A, B> =
  (<T>() => T extends A ? 1 : 2) extends <T>() => T extends B ? 1 : 2
    ? true
    : false;

// This will error if R is not exactly number

type _Test = Assert<IsEqual<R, number>>;
```

`npm run compile`

### Expected Result

Success

### Actual Result

```
index.ts:15:21 - error TS2344: Type 'false' does not satisfy the constraint 'true'.

15 type _Test = Assert<IsEqual<R, number>>;
                       ~~~~~~~~~~~~~~~~~~
```

### Additional Notes

The IDE experience is likewise unusable

<img width="493" height="109" alt="Image" src="https://github.com/user-attachments/assets/40cfb35f-decf-4b57-af4c-ac3a18bd7f1d" />

### Cause

https://github.com/praha-inc/byethrow/blob/main/tsconfig.json#L8

Using the `"bundler"` type of module resolution allows for modules to elide file extensions. Changing this to `nodenext` will maintain compatibility with browsers/bundlers and extend compat to nodejs esm projects.

--- a/.changeset/five-crabs-ask.md
+++ b/.changeset/five-crabs-ask.md
@@ -1,0 +1,5 @@
+---
+"@praha/byethrow": patch
+---
+
+Add file extensions to import paths in TypeScript declaration files

--- a/packages/byethrow/rslib.config.ts
+++ b/packages/byethrow/rslib.config.ts
@@ -1,5 +1,23 @@
 import { defineConfig } from '@rslib/core';
 
+import type { Format, LibConfig } from '@rslib/core';
+
+const createLibrary = (format: Format): LibConfig => ({
+  format,
+  bundle: false,
+  dts: true,
+  redirect: {
+    dts: {
+      extension: true,
+    },
+  },
+  output: {
+    distPath: {
+      root: `./dist/${format}`,
+    },
+  },
+});
+
 export default defineConfig({
   source: {
     entry: {
@@ -12,25 +30,7 @@ export default defineConfig({
     tsconfigPath: './tsconfig.build.json',
   },
   lib: [
-    {
-      format: 'cjs',
-      bundle: false,
-      dts: true,
-      output: {
-        distPath: {
-          root: './dist/cjs',
-        },
-      },
-    },
-    {
-      format: 'esm',
-      bundle: false,
-      dts: true,
-      output: {
-        distPath: {
-          root: './dist/esm',
-        },
-      },
-    },
+    createLibrary('cjs'),
+    createLibrary('esm'),
   ],
 });

